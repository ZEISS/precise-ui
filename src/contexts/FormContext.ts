import * as React from 'react';

export interface FormValueNotifier {
  state: {
    value: any;
  };
  props: {
    name?: string;
  };
  setState(state: { value: any }): void;
}

export interface FormValueChange {
  name: string;
  value: any;
}

export interface FormContextType {
  change(data: FormValueChange): void;
  subscribe(field: FormValueNotifier): void;
  unsubscribe(field: FormValueNotifier): void;
}

export const FormContext = React.createContext<FormContextType | undefined>(undefined);
