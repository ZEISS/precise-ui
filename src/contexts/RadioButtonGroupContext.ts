import * as React from 'react';

export interface RadioButtonGroupNotifier {
  setValue(value: boolean): void;
  name?: string;
}

export interface RadioButtonGroupContextType {
  select(rb: RadioButtonGroupNotifier): void;
  subscribe(rb: RadioButtonGroupNotifier): void;
  unsubscribe(rb: RadioButtonGroupNotifier): void;
}

export const RadioButtonGroupContext = React.createContext<RadioButtonGroupContextType | undefined>(undefined);
