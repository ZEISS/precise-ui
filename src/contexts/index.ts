export * from './RadioButtonGroupContext';
export * from './FormContext';
