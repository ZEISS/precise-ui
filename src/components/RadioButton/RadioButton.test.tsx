import * as React from 'react';
import * as enzyme from 'enzyme';
import { RadioButton } from './';
import { RadioButtonGroupContextType, RadioButtonGroupNotifier } from '../../contexts';

function getRadioButtonWithContext(radioGroupCtx?: RadioButtonGroupContextType) {
  jest.doMock('../../contexts', () => ({
    RadioButtonGroupContext: {
      Consumer: props => props.children(radioGroupCtx),
    },
    FormContext: {
      Consumer: props => props.children(),
    },
  }));

  return require('./').RadioButton;
}

beforeEach(() => {
  jest.resetModules();
});

describe('<RadioButton />', () => {
  it('should render the disabled, uncontrolled RadioButton', () => {
    const wrapper = enzyme.shallow(<RadioButton disabled defaultValue />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the controlled RadioButton with a label', () => {
    const wrapper = enzyme.shallow(<RadioButton value>Value</RadioButton>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should subscribe for received context', () => {
    let subscribed = false;
    const radioGroupCtx = {
      subscribe(radio: RadioButtonGroupNotifier) {
        if (radio) {
          subscribed = true;
        }
      },
      unsubscribe() {},
      select() {},
    };
    const RadioButtonWithContext = getRadioButtonWithContext(radioGroupCtx);
    const wrapper = enzyme.mount(<RadioButtonWithContext>Value</RadioButtonWithContext>);
    expect(subscribed).toEqual(true);
  });

  it('should skip context subscribe in controlled mode (value set)', () => {
    let subscribed = false;
    const radioGroupCtx = {
      subscribe(radio: RadioButtonGroupNotifier) {
        subscribed = true;
      },
      unsubscribe() {},
      select() {},
    };
    const RadioButtonWithContext = getRadioButtonWithContext(radioGroupCtx);
    const wrapper = enzyme.mount(<RadioButtonWithContext value>Value</RadioButtonWithContext>);
    expect(subscribed).toEqual(false);
  });

  it('should unsubscribe from context when unmounted', () => {
    let unsubscribed = false;
    const radioGroupCtx = {
      unsubscribe(radio: RadioButtonGroupNotifier) {
        if (radio) {
          unsubscribed = true;
        }
      },
      subscribe() {},
      select() {},
    };
    const RadioButtonWithContext = getRadioButtonWithContext(radioGroupCtx);
    const wrapper = enzyme.mount(<RadioButtonWithContext>Value</RadioButtonWithContext>);
    wrapper.unmount();
    expect(unsubscribed).toEqual(true);
  });

  it('should skip unsubscribing from context when unmounted while in controlled mode (value set)', () => {
    let unsubscribed = false;
    const radioGroupCtx = {
      unsubscribe(radio: RadioButtonGroupNotifier) {
        unsubscribed = true;
      },
      subscribe() {},
      select() {},
    };
    const RadioButtonWithContext = getRadioButtonWithContext(radioGroupCtx);
    const wrapper = enzyme.mount(<RadioButtonWithContext value>Value</RadioButtonWithContext>);
    wrapper.unmount();
    expect(unsubscribed).toEqual(false);
  });

  it('should notify context when selected', () => {
    let selected = false;
    const radioGroupCtx = {
      select(radio: RadioButtonGroupNotifier) {
        if (radio) {
          selected = true;
        }
      },
      subscribe() {},
      unsubscribe() {},
    };
    const RadioButtonWithContext = getRadioButtonWithContext(radioGroupCtx);
    const wrapper = enzyme.mount(<RadioButtonWithContext name="radio">Value</RadioButtonWithContext>);
    wrapper
      .find({ name: 'radio' })
      .first()
      .simulate('click');
    expect(selected).toEqual(true);
  });

  it('should not notify context when trying to select disabled RadioButton', () => {
    let selected = false;
    const radioGroupCtx = {
      select(radio: RadioButtonGroupNotifier) {
        if (radio) {
          selected = true;
        }
      },
      subscribe() {},
      unsubscribe() {},
    };
    const RadioButtonWithContext = getRadioButtonWithContext(radioGroupCtx);
    const wrapper = enzyme.mount(
      <RadioButtonWithContext disabled name="radio">
        Value
      </RadioButtonWithContext>,
    );
    wrapper
      .find({ name: 'radio' })
      .first()
      .simulate('click');
    expect(selected).toEqual(false);
  });
});
