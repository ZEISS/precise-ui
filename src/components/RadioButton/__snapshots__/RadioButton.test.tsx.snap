// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton /> should render the controlled RadioButton with a label 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`<RadioButton /> should render the disabled, uncontrolled RadioButton 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
